### DISCUSSION TITLE
Drop principles related to mutable data

### DISCUSSION THREAD

**Discussion by magnus-madsen** [flix/flix#9041] [General]
We have the following principles:

![CleanShot 2024-10-28 at 12 34 14](https://github.com/user-attachments/assets/1b88d094-e11c-4fd9-a3a7-1640c48242f9)

I would like to drop them. In particular, I would:

- Drop the idea that destructive operations are marked with exclamation point. A much better solution is to use the effect system.
- Ensure that all operations are in-place (e.g. `reverse` would be in place).
- We can still use a mutable data structure as immutable ones. For example, one can call `Foldable.whatever`.

**Comment by magnus-madsen** [comment]
@mlutze @JonathanStarup @jaschdoc @stephentetley I would like to know your thoughts.

**Reply by JonathanStarup** [reply]
You could still write `inPlaceReverse` and `reverse` for array and they would have the same effect, so we need a convention to not confuse the two. One solution is the `!` for in place functions but we could also never have the "functional" ones, requiring users to write `arr |> copy |> reverse`

**Reply by magnus-madsen** [reply]
I think we omit the functional ones and do as you say with copy.

**Reply by JonathanStarup** [reply]
I think there is still confusing cases though. This rule would imply that `map` is in place and `transform` doesn't exist? what about `slice` fx. It cannot be in place but if the programmer doesn't know that it is a surprise. Should `scan` be in place (it also cant, unless we reduce generality)?

**Reply by stephentetley** [reply]
I'd be very happy to see the `!` suffix go.

**Reply by jaschdoc** [reply]
I think it is fine as long as destructive operations return `Unit` so it is clear that whatever happened must have happened to the mutable data. That way functional code breaks (which is good in this case) if it is changed to imperative style, making it impossible to accidentally confuse semantics.

E.g.
```flix
def addOne(l: List[Int32]): List[Int32] = List.map(x -> x + 1, l)
```
becomes

```flix
def addOne(a: Array[Int32, r]): Unit \ r = Array.map(x -> x + 1, a)
```
